src/Connection/Internal/Http2ConnectionProcessor.php

```diff
@@ -591,17 +591,20 @@ static function () {
                 $this->releaseStream($streamId, $exception);
             });
 
+            $promise = $stream->pendingResponse->promise();
+            $promise->onResolve(function () use ($cancellationToken, $cancellationId): void {
+                $cancellationToken->unsubscribe($cancellationId);
+            });
```

Comment:
This isn't correct either, I guess it has to subscribe to the `$response->getTrailers()` promise.

Reply:
Yep. This keeps getting more complicated. 😛

Reply:
Maybe we should just keep the current behavior?

Reply:
I'm thinking so. The user can return a promise that isn't resolved until they know they want to receive the push.

@nicolas-grekas, you return a deferred promise from the handler, subsequently failing the deferred to cancel the push.

Reply:
Got it, thank you and sorry for the noise, I should have thought about deferred earlier.